# Project Overview
## E-Commerce Sales & Customer Analysis

### Project Overview
- This project analyzes e-commerce sales and customer data to identify sales trends, customer behavior, category performance, and top-performing products.

### Objectives

- Clean and validate the raw e-commerce dataset
- Explore sales and customer patterns
- Analyze monthly sales trends
- Identify high-value customer segments
- Evaluate category and product performance
- Generate business insights for decision-making

### Tools

- Python
- Pandas
- NumPy
- Matplotlib
- Seaborn

# 01_Import Libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# 02_Load Dataset

In [3]:
df = pd.read_csv("../data/ecommerce_sales_raw.csv")

# 03_Data Inspection

In [4]:
df.head()

,OrderID,CustomerID,OrderDate,ProductID,Quantity,Discount,PaymentMethod,Status,Age,City,SignupDate,CustomerSegment,ProductName,Category,UnitPrice,Sales,OrderValue
0,500001,103695,28/08/2025,2003,4,10,Gateway,Completed,36,Qom,05/03/2024,Regular,USB-C Cable,Accessories,9,32,32
1,500002,107935,31/05/2024,2014,1,10,Wallet,Completed,49,Kerman,04/06/2025,Regular,Backpack,Accessories,42,38,38
2,500003,105141,10/08/2025,2002,1,20,CardToCard,Completed,36,Tehran,19/10/2025,Regular,Mechanical Keyboard,Electronics,62,50,50
3,500004,108780,11/10/2024,2012,1,10,Gateway,Completed,45,Shiraz,07/12/2023,Regular,Office Chair,Home Office,180,162,162
4,500005,101187,19/02/2024,2008,2,0,Gateway,Completed,24,Karaj,02/12/2023,Regular,Phone Case,Accessories,14,28,28


In [5]:
df.shape

(49222, 17)

In [6]:
df.columns

Index(['OrderID', 'CustomerID', 'OrderDate', 'ProductID', 'Quantity',
       'Discount', 'PaymentMethod', 'Status', 'Age', 'City', 'SignupDate',
       'CustomerSegment', 'ProductName', 'Category', 'UnitPrice', 'Sales',
       'OrderValue'],
      dtype='object')

In [7]:
df.info()   #Dataset information

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49222 entries, 0 to 49221
Data columns (total 17 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   OrderID          49222 non-null  int64 
 1   CustomerID       49222 non-null  int64 
 2   OrderDate        49222 non-null  object
 3   ProductID        49222 non-null  int64 
 4   Quantity         49222 non-null  int64 
 5   Discount         49222 non-null  int64 
 6   PaymentMethod    49222 non-null  object
 7   Status           49222 non-null  object
 8   Age              49222 non-null  int64 
 9   City             49222 non-null  object
 10  SignupDate       49222 non-null  object
 11  CustomerSegment  49222 non-null  object
 12  ProductName      49222 non-null  object
 13  Category         49222 non-null  object
 14  UnitPrice        49222 non-null  int64 
 15  Sales            49222 non-null  int64 
 16  OrderValue       49222 non-null  int64 
dtypes: int64(9), object(8)
memory u

In [8]:
df.isnull().sum()   #Missing values in the dataset

OrderID            0
CustomerID         0
OrderDate          0
ProductID          0
Quantity           0
Discount           0
PaymentMethod      0
Status             0
Age                0
City               0
SignupDate         0
CustomerSegment    0
ProductName        0
Category           0
UnitPrice          0
Sales              0
OrderValue         0
dtype: int64

In [9]:
df.duplicated().sum()    #checking for duplicate values

np.int64(0)

In [10]:
df.describe()  #Statistical summary of the dataset

,OrderID,CustomerID,ProductID,Quantity,Discount,Age,UnitPrice,Sales,OrderValue
count,49222.000000,49222.000000,49222.000000,49222.000000,49222.000000,49222.000000,49222.000000,49222.000000,49222.000000
mean,525007.608366,105002.304356,2009.460160,1.881923,7.485576,41.283369,40.286579,69.985576,69.985576
std,14435.345449,2888.912895,5.578603,1.079382,7.754378,13.681355,42.613062,94.483273,94.483273
min,500001.000000,100001.000000,2001.000000,-2.000000,0.000000,18.000000,7.000000,-170.000000,-170.000000
25%,512507.250000,102504.000000,2004.000000,1.000000,0.000000,29.000000,11.000000,17.000000,17.000000
50%,525016.500000,105005.000000,2008.000000,2.000000,5.000000,41.000000,28.000000,38.000000,38.000000
75%,537506.750000,107497.000000,2014.000000,2.000000,10.000000,53.000000,58.000000,84.000000,84.000000
max,550000.000000,110000.000000,2020.000000,5.000000,30.000000,65.000000,260.000000,1300.000000,1300.000000


In [11]:
df.nunique()    #Number of unique values in each column

OrderID            49222
CustomerID          9820
OrderDate            912
ProductID             20
Quantity               8
Discount               6
PaymentMethod          4
Status                 3
Age                   48
City                  10
SignupDate          1094
CustomerSegment        3
ProductName           20
Category               6
UnitPrice             20
Sales                290
OrderValue           290
dtype: int64

# 04_Data Cleaning

In [12]:
df_clean = df.copy()

df_clean = df_clean[
    (df_clean["Quantity"] > 0) &
    (df_clean["Sales"] >= 0) &
    (df_clean["OrderValue"] >= 0)
].copy()

In [13]:
# Converting date columns to datetime format

df_clean["OrderDate"] = pd.to_datetime(
    df_clean["OrderDate"],
    format="%d/%m/%Y"
)

df_clean["SignupDate"] = pd.to_datetime(
    df_clean["SignupDate"],
    format="%d/%m/%Y"
)

In [14]:
df_clean.shape

(49198, 17)

# 05_Data Validation

In [15]:
df_clean.isnull().sum()   #Checking for missing values after cleaning

OrderID            0
CustomerID         0
OrderDate          0
ProductID          0
Quantity           0
Discount           0
PaymentMethod      0
Status             0
Age                0
City               0
SignupDate         0
CustomerSegment    0
ProductName        0
Category           0
UnitPrice          0
Sales              0
OrderValue         0
dtype: int64

In [16]:
(df_clean["Quantity"] <= 0).sum()  #Checking for negative values in Quantity column

np.int64(0)

In [17]:
(df_clean["Sales"] < 0).sum()  #Checking for negative values in Sales column

np.int64(0)

In [18]:
# Calculating expected sales

expected_sales = (
    df_clean["Quantity"]
    * df_clean["UnitPrice"] * (1 - df_clean["Discount"] /100)
).round(0)
(df_clean["Sales"] == expected_sales).mean()  

np.float64(1.0)

In [19]:
# Saving the cleaned dataset to a new CSV file
df_clean.to_csv("../data/E-Commerce_Sales_Cleaned.csv", index=False)

# 06_Feature Engineering

In [20]:
# Creating new columns for Year, Month, MonthName, Quarter, and YearMonth

df_clean["Year"] = df_clean["OrderDate"].dt.year
df_clean["Month"] = df_clean["OrderDate"].dt.month
df_clean["MonthName"] = df_clean["OrderDate"].dt.month_name()
df_clean["Quarter"] = "Q" + df_clean["OrderDate"].dt.quarter.astype(str)
df_clean["YearMonth"] = df_clean["OrderDate"].dt.to_period("M")

In [21]:
# Creating a new column for CustomerType based on the number of orders per customer

orders_per_customer = df_clean.groupby("CustomerID").size()
df_clean["OrderCount"] = df_clean["CustomerID"].map(orders_per_customer)

df_clean["CustomerType"] = np.where(
    df_clean["OrderCount"] == 1,
    "New Customer",
    "Repeat Customer"
)

df_clean["CustomerType"].value_counts()

CustomerType
Repeat Customer    48878
New Customer         320
Name: count, dtype: int64

# 07_Customer Analysis

In [22]:
# Creating a customer summary DataFrame with aggregated metrics

customer_summary = (
    df_clean.groupby("CustomerID")
    .agg(
        OrderCount=("OrderID", "count"),
        TotalSales=("Sales", "sum"),
        AvgOrderValue=("Sales", "mean")
    )
    .reset_index()
)

customer_summary["CustomerType"] = np.where(
    customer_summary["OrderCount"] == 1,
    "New Customer",
    "Repeat Customer"
)

In [23]:
customer_summary["CustomerType"].value_counts() # Count of each customer type

CustomerType
Repeat Customer    9500
New Customer        320
Name: count, dtype: int64

In [24]:
# Creating value segments based on total sales

customer_summary["CustomerValueSegment"] = pd.cut(
    customer_summary["TotalSales"],
    bins=[0, 166, 290, 465, float("inf")],
    labels=["Low Value", "Medium-Low", "Medium-High", "High Value"],
    include_lowest=True
)

customer_summary["CustomerValueSegment"].value_counts()


CustomerValueSegment
Low Value      2460
Medium-Low     2456
High Value     2454
Medium-High    2450
Name: count, dtype: int64

In [25]:
# Grouping by CustomerValueSegment and calculating the mean of OrderCount, TotalSales, and AvgOrderValue

customer_summary.groupby(
    "CustomerValueSegment",
    observed=True
)[
    ["OrderCount", "TotalSales", "AvgOrderValue"]
].mean()

,OrderCount,TotalSales,AvgOrderValue
CustomerValueSegment,,,
Low Value,3.077642,96.589024,34.054034
Medium-Low,4.614821,227.031352,55.870133
Medium-High,5.628571,370.826939,73.922417
High Value,6.724939,709.880196,117.246639


In [27]:
# Grouping by CustomerSegment and calculating the number of customers, total sales, average order value, and number of orders

customer_segment_analysis = (
    df_clean.groupby("CustomerSegment")
    .agg(
        Customers=("CustomerID", "nunique"),
        TotalSales=("Sales", "sum"),
        AvgOrderValue=("Sales", "mean"),
        Orders=("OrderID", "count")
    )
    .sort_values("TotalSales", ascending=False)
)

customer_segment_analysis

,Customers,TotalSales,AvgOrderValue,Orders
CustomerSegment,,,,
Regular,5468,1915222,70.200938,27282
New,3399,1196523,70.181418,17049
VIP,953,334025,68.630573,4867


# 08_Sales Trend Analysis

In [28]:
# Grouping by YearMonth and calculating total sales, number of orders, and number of unique customers

monthly_sales = (
    df_clean.groupby("YearMonth")
    .agg(
        TotalSales=("Sales", "sum"),
        Orders=("OrderID", "count"),
        Customers=("CustomerID", "nunique")
    )
    .reset_index()
)

monthly_sales["MoM_Growth"] = monthly_sales["TotalSales"].pct_change() * 100
monthly_sales

,YearMonth,TotalSales,Orders,Customers,MoM_Growth
0,2024-01,121921,1747,1594,NaN
1,2024-02,114664,1592,1464,-5.952215
2,2024-03,128150,1815,1651,11.761320
3,2024-04,115950,1675,1549,-9.520094
4,2024-05,123700,1742,1600,6.683915
5,2024-06,116784,1690,1562,-5.590946
6,2024-07,128689,1737,1600,10.194033
7,2024-08,129987,1813,1662,1.008633
8,2024-09,116032,1643,1499,-10.735689
9,2024-10,117260,1730,1589,1.058329


In [29]:
# Find the months with the highest and lowest sales

max_sales = monthly_sales.loc[monthly_sales["TotalSales"].idxmax()]
min_sales = monthly_sales.loc[monthly_sales["TotalSales"].idxmin()]

max_sales, min_sales

(YearMonth       2025-05
 TotalSales       130625
 Orders             1765
 Customers          1623
 MoM_Growth    13.287484
 Name: 16, dtype: object,
 YearMonth       2026-02
 TotalSales        75132
 Orders             1074
 Customers          1025
 MoM_Growth   -42.449636
 Name: 25, dtype: object)

# 09_Category Analysis

In [30]:
# Grouping by Category and calculating total sales, number of orders, number of unique customers, and average order value

category_analysis = (
    df_clean.groupby("Category")
    .agg(
        TotalSales=("Sales", "sum"),
        Orders=("OrderID", "count"),
        Customers=("CustomerID", "nunique"),
        AvgOrderValue=("Sales", "mean")
    )
    .reset_index()
    .sort_values("TotalSales", ascending=False)
)

category_analysis

,Category,TotalSales,Orders,Customers,AvgOrderValue
1,Electronics,1751491,17758,8235,98.631096
0,Accessories,553178,18180,8297,30.427833
3,Home Office,446065,3121,2676,142.923742
5,Wearables,429497,3041,2607,141.235449
2,Gaming,198302,1661,1520,119.387116
4,Stationery,67237,5437,4181,12.366562


# 10_Product Analysis

In [31]:
# Grouping by ProductID and ProductName and calculating total sales, number of orders, and quantity sold

product_analysis = (
    df_clean.groupby(["ProductID", "ProductName"])
    .agg(
        TotalSales=("Sales", "sum"),
        Orders=("OrderID", "count"),
        Quantity=("Quantity", "sum")
    )
    .reset_index()
    .sort_values("TotalSales", ascending=False)
)

product_analysis.head(10)

,ProductID,ProductName,TotalSales,Orders,Quantity
6,2007,Headphones,333531,2558,4804
11,2012,Office Chair,329292,1032,1977
18,2019,Tablet,292344,658,1228
8,2009,Smart Watch,259590,1251,2344
17,2018,Microphone,230309,1538,2931
1,2002,Mechanical Keyboard,229676,2152,3995
14,2015,Bluetooth Speaker,213427,2131,3983
19,2020,Gaming Controller,198302,1661,3156
3,2004,Power Bank,180004,2970,5560
9,2010,Fitness Band,169907,1790,3332


# 11_Key Insights
  ### Sales Performance

  - May 2025 recorded the highest monthly sales at $130.6K.
  - February 2026 recorded the lowest monthly sales at $75.1K.
  - Sales declined 42.45% MoM in February 2026, followed by a partial recovery in March.

  ### Customer Insights

  - Repeat customers represent the majority of observed customers.
  - High Value customers have higher average order frequency and spending than lower-value customer segments.
  - Regular customers generate the highest total sales among the dataset's predefined customer segments.

  ### Category Insights

  - Electronics is the largest sales-generating category.
  - Accessories has the highest order volume but a relatively low average order value.
  - Home Office has the highest average order value despite lower order volume.

  ### Product Insights

  - Headphones is the top product by total sales.
  - Office Chair generates nearly the same sales as Headphones with fewer orders, indicating higher sales value per order.

# 12_Data Limitations

- The dataset covers January 2024 to June 2026.
- 2026 contains only six months of data, so full-year comparisons with 2024 and 2025 should be interpreted with caution.
- CustomerValueSegment is based on quartiles of observed customer sales, so the segment thresholds are relative to this dataset rather than business-defined thresholds.
- The dataset does not contain a Profit field, so profitability analysis is outside the scope of this project.

# 13_Export Cleaned Dataset

In [ ]:
# Saving the cleaned dataset to a new CSV file

df_clean.to_csv(
    "../data/E-Commerce_Sales_Cleaned.csv", index=False
)